src/components/contex/TimerContext.tsx

```diff
@@ -0,0 +1,134 @@
+import React, { createContext, useReducer, useContext, ReactNode } from "react";
+
+interface Timer {
+  id: number;
+  type: string;
+  duration: number;
+  state: "not running" | "running" | "completed";
+  config?: {
+    rounds?: number;
+    workTime?: number;
+    restTime?: number;
+    timePerRound?: number;  // property for XY timer
+  };
+}
+
+interface TimerState {
+  timers: Timer[];
+  activeTimerIndex: number | null;
+  isRunning: boolean;
+  totalTime: number;
+}
+
+type TimerAction =
+  | { type: "ADD_TIMER"; payload: Timer }
+  | { type: "REMOVE_TIMER"; payload: number }
+  | { type: "START_TIMER"; payload: number }
+  | { type: "COMPLETE_TIMER"; payload: number }
+  | { type: "RESET_WORKOUT" }
+  | { type: "FAST_FORWARD" }
+  | { type: "PAUSE_RESUME_WORKOUT" }
+  | { type: "UPDATE_TOTAL_TIME" };
+
+const initialState: TimerState = {
+  timers: [],
+  activeTimerIndex: null,
+  isRunning: false,
+  totalTime: 0,
+};
+
+function timerReducer(state: TimerState, action: TimerAction): TimerState {
+  switch (action.type) {
+    case "ADD_TIMER":
+      return {
+        ...state,
+        timers: [...state.timers, action.payload],
+        totalTime: state.totalTime + action.payload.duration,
+      };
+    
+    case "REMOVE_TIMER":
+      return {
+        ...state,
+        timers: state.timers.filter((_, idx) => idx !== action.payload),
+        totalTime: state.timers.reduce((sum, timer, idx) => 
+          idx !== action.payload ? sum + timer.duration : sum, 0
+        ),
+      };
+    
+    case "START_TIMER":
+      return {
+        ...state,
+        timers: state.timers.map((timer, idx) => ({
+          ...timer,
+          state: idx === action.payload ? "running" : timer.state,
+        })),
+        activeTimerIndex: action.payload,
+        isRunning: true,
+      };
+    
+    case "COMPLETE_TIMER":
+      const nextIndex = action.payload + 1 < state.timers.length ? action.payload + 1 : null;
+      return {
+        ...state,
+        timers: state.timers.map((timer, idx) => ({
+          ...timer,
+          state: idx === action.payload ? "completed" : timer.state,
+        })),
+        activeTimerIndex: nextIndex,
+        isRunning: nextIndex !== null,
+      };
+    
+    case "RESET_WORKOUT":
+      return {
+        ...state,
+        timers: state.timers.map(timer => ({
+          ...timer,
+          state: "not running",
+        })),
+        activeTimerIndex: null,
+        isRunning: false,
+      };
+    
+    case "FAST_FORWARD":
+      if (state.activeTimerIndex === null) return state;
+      const newIndex = state.activeTimerIndex + 1;
+      return {
+        ...state,
+        activeTimerIndex: newIndex < state.timers.length ? newIndex : null,
+        isRunning: newIndex < state.timers.length,
+        timers: state.timers.map((timer, idx) => ({
+          ...timer,
+          state: idx === state.activeTimerIndex ? "completed" :
```

Comment:
Yea I didn't like this at all and I honestly didn't know how to go about refactoring this.

Reply:
Try to keep in mind the idea of single responsibility - it's a really good guiding principle for software engineering. Every module/function/component should ideally be responsible for one thing only (obviously not always possible). If your function is long, or you have a bunch of if/then statements that are nested, or you see nested ternary operators like this - this should be a code smell that something needs to be re-factored because your function is responsible for too much.

Two good approaches to starting the re-factor - top-down and bottom-up. Top-down is beginning with a holistic view: Does the overall logic make sense? Is there anything I can simplify? Can I draw out (on paper, on a whiteboard) the logic of the different possible states that my application can be in? Nothing fancy here, but just drawing this out can be a huge help. 

Bottom-up: Without thinking about the whole context, try to move code to a separate function. Don't worry about making the most compact code - make it readable. Maybe this:
```
        timers: state.timers.map((timer, idx) => ({
          ...timer,
          state: idx === state.activeTimerIndex ? "completed" :      idx === newIndex ? "running" : timer.state,
        })),
      };
```
becomes
```
        timers: setTimersFastForwardState({
              timers: state.timers, 
              activeTimerIndex: state.activeTimerIndex, 
              nextTimerIndex: newIndex
        })
```
and elsewhere, you might have the function 
```
const setTimersFastForwardState = {timers, activeTimerIndex, nextTimerIndex: newIndex}) => {
    return timers.map((timer, idx) => {
          const isNextTimer = idx === newIndex
          const isCurrentTimer =  idx === state.activeTimerIndex 

          if (!isNextTimer && !isCurrentTimer) return timer;

         return {
           ...timer,
          state: isCurrentTimer ? "stopped" : "running"
        }
    }
}
```
More verbose, but also more readable and it's kind of self-documenting: it's a bit easier to understand what `idx === state.activeTimerIndex` is when you store it in a variable called `isCurrentTimer`

Reply:
Or maybe you feel like that's still doing too much, so you want to re-factor again: 
```
        timers: state.timers.map((timer, idx) => ({
            ...timer,
           state: setTimerFastForwardState({
               timer,
               currentTimerIndex: idx,
               activeTimerIndex: state.activeTimerIndex, 
               nextTimerIndex: newIndex
           })
```
and elsewhere you have:
```
const setTimerFastForwardState = ({timer, currentTimerIndex, activeTimerIndex, nextTimerIndex}) => {
   if (currentTimerIndex === activeTimerIndex) return "stopped"
   if (currentTimerIndex === nextTimerIndex) return "running"
   return timer.state

}
```